clients/src/test/java/org/apache/kafka/common/utils/LogCaptureAppender.java

```diff
@@ -74,29 +80,51 @@ public Optional<String> getThrowableClassName() {
         }
     }
 
+    public LogCaptureAppender(String name) {
+        super(name, null, null, true, Property.EMPTY_ARRAY);
+    }
+
     public static LogCaptureAppender createAndRegister() {
-        final LogCaptureAppender logCaptureAppender = new LogCaptureAppender();
-        Logger.getRootLogger().addAppender(logCaptureAppender);
+        final LogCaptureAppender logCaptureAppender = new LogCaptureAppender("LogCaptureAppender");
+        Logger logger = LogManager.getRootLogger();
+        logCaptureAppender.addToLogger(logger);
         return logCaptureAppender;
     }
 
     public static LogCaptureAppender createAndRegister(final Class<?> clazz) {
-        final LogCaptureAppender logCaptureAppender = new LogCaptureAppender();
-        Logger.getLogger(clazz).addAppender(logCaptureAppender);
+        final LogCaptureAppender logCaptureAppender = new LogCaptureAppender("LogCaptureAppender");
+        Logger logger = LogManager.getLogger(clazz);
+        logCaptureAppender.addToLogger(logger);
         return logCaptureAppender;
     }
 
-    public void setClassLogger(final Class<?> clazz, Level level) {
-        logLevelChanges.add(new LogLevelChange(Logger.getLogger(clazz).getLevel(), clazz));
-        Logger.getLogger(clazz).setLevel(level);
+    public void addToLogger(Logger logger) {
+        org.apache.logging.log4j.core.Logger coreLogger = (org.apache.logging.log4j.core.Logger) logger;
+        this.start();
+        coreLogger.addAppender(this);
+        loggers.add(coreLogger);
     }
 
-    public static void unregister(final LogCaptureAppender logCaptureAppender) {
-        Logger.getRootLogger().removeAppender(logCaptureAppender);
+    public void setClassLogger(final Class<?> clazz, Level level) {
+        LoggerContext ctx = (LoggerContext) LogManager.getContext(false);
+        Configuration config = ctx.getConfiguration();
+        String loggerName = clazz.getName();
+        LoggerConfig loggerConfig = config.getLoggerConfig(loggerName);
+
+        Level originalLevel = loggerConfig.getLevel();
+        logLevelChanges.add(new LogLevelChange(originalLevel, clazz));
+
+        if (!loggerConfig.getName().equals(loggerName)) {
+            LoggerConfig newLoggerConfig = new LoggerConfig(loggerName, level, true);
+            config.addLogger(loggerName, newLoggerConfig);
+        } else {
+            loggerConfig.setLevel(level);
+        }
+        ctx.updateLoggers();
     }
 
     @Override
-    protected void append(final LoggingEvent event) {
+    public void append(final LogEvent event) {
         synchronized (events) {
             events.add(event);
```

Comment:
```suggestion
    public void append(final LogEvent event) {
        synchronized (events) {
            events.add(event.toImmutable());
```
By default log events are mutable and bound the one thread. They are cleared as soon as the logging call exits.
There is a [`log4j2.enableThreadlocals` kill switch](https://logging.apache.org/log4j/2.x/manual/systemproperties.html#log4j2.enableThreadlocals) that you can store in a [`log4j2.component.properties`](https://logging.apache.org/log4j/2.x/manual/systemproperties.html#property-sources) on the test classpath, but it is easier to just take an immutable snapshot.

You can also replace `LogCaptureAppender` with [`ListAppender`](https://github.com/apache/logging-log4j2/blob/2.x/log4j-core-test/src/main/java/org/apache/logging/log4j/core/test/appender/ListAppender.java) from the [`log4j-core-test` artifact](https://logging.apache.org/log4j/2.x/components.html#log4j-core-test). You can set it up with a config like:

```xml
<Configuration>
  <Appenders>
    <List name="LIST"/>
  </Appenders>
  <Loggers>
    <Root level="DEBUG">
      <AppenderRef ref="LIST"/>
    </Root>
  </Loggers>
</Configuration>

Reply:
Very appreciate!
I have applied it and tested it locally; it works like a charm.
As for replacing `LogCaptureAppender`，I think it's a great idea. 
IMHO, leveraging an existing tool is better than building our own. 
I will file a JIRA to initiate a discussion on this.

Reply:
https://issues.apache.org/jira/browse/KAFKA-17963